src/dbnode/persist/fs/merger.go

```diff
@@ -0,0 +1,293 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"io"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/digest"
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/dbnode/storage/index/convert"
+	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/dbnode/x/xio"
+	"github.com/m3db/m3/src/x/checked"
+	"github.com/m3db/m3/src/x/context"
+	"github.com/m3db/m3/src/x/ident"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+type merger struct {
+	reader         DataFileSetReader
+	blockAllocSize int
+	srPool         xio.SegmentReaderPool
+	multiIterPool  encoding.MultiReaderIteratorPool
+	identPool      ident.Pool
+	encoderPool    encoding.EncoderPool
+	nsOpts         namespace.Options
+}
+
+// NewMerger returns a new Merger. This implementation is in charge of merging
+// the data from an existing fileset with a merge target, giving precendence
+// to the data from the merge target. This merged data is then persisted.
+//
+// Note that the merger does not know how or where this merged data is
+// persisted since it just uses the flushPreparer that is passed in. Further,
+// it does not signal to the database of the existence of the newly persisted
+// data, nor does it clean up the original fileset.
+func NewMerger(
+	reader DataFileSetReader,
+	blockAllocSize int,
+	srPool xio.SegmentReaderPool,
+	multiIterPool encoding.MultiReaderIteratorPool,
+	identPool ident.Pool,
+	encoderPool encoding.EncoderPool,
+	nsOpts namespace.Options,
+) Merger {
+	return &merger{
+		reader:         reader,
+		blockAllocSize: blockAllocSize,
+		srPool:         srPool,
+		multiIterPool:  multiIterPool,
+		identPool:      identPool,
+		encoderPool:    encoderPool,
+		nsOpts:         nsOpts,
+	}
+}
+
+// Merge merges data from a fileset with a merge target and persists it.
+func (m *merger) Merge(
```

Comment:
Can you add to the doc string to say that the caller is responsible for finalizing all data structures in the `MergeWith`? Alternatively you could make it so the `MergeWith` has a `Finalize()` method that the merger can call, and then if the caller doesn't want their data finalized they can just make that operation a no-op. 

Either way is fine to me

Reply:
Sure, I'll go with the former.